src/borg/helpers/shellpattern.py

```diff
@@ -60,7 +60,82 @@ def translate(pat, match_end=r"\Z"):
                 elif stuff[0] == "^":
                     stuff = "\\" + stuff
                 res += "[%s]" % stuff
+        elif c in "(|)":
+            if i > 0 and pat[i - 1] != "\\":
+                res += c
         else:
             res += re.escape(c)
 
     return "(?ms)" + res + match_end
+
+
+def _parse_braces(pat):
+    """Returns the index values of paired braces in `pat` as a list of tuples.
+
+    The dict's keys are the indexes corresponding to opening braces. Initially,
+    they are set to a value of `None`. Once a corresponding closing brace is found,
+    the value is updated. All dict keys with a positive int value are valid pairs.
+
+    Cannot rely on re.match("[^\\(\\\\)*]?{.*[^\\(\\\\)*]}") because, while it
+    does handle unpaired braces and nested pairs of braces, it misses sequences
+    of paired braces. E.g.: "{foo,bar}{bar,baz}" would translate, incorrectly, to
+    "(foo|bar\\}\\{bar|baz)" instead of, correctly, to "(foo|bar)(bar|baz)"
+
+    So this function parses in a left-to-right fashion, tracking pairs with a LIFO
+    queue: pushing opening braces on and popping them off when finding a closing
+    brace.
+    """
+    curly_q = LifoQueue()
+    pairs: dict[int, int] = dict()
+
+    for idx, c in enumerate(pat):
+        if c == "{":
+            if idx == 0 or pat[idx - 1] != "\\":
+                # Opening brace is not escaped.
+                # Add to dict
+                pairs[idx] = None
+                # Add to queue
+                curly_q.put(idx)
+        if c == "}" and curly_q.qsize():
+            # If queue is empty, then cannot close pair.
+            if idx > 0 and pat[idx - 1] != "\\":
+                # Closing brace is not escaped.
+                # Pop off the index of the corresponding opening brace, which
+                # provides the key in the dict of pairs, and set its value.
+                pairs[curly_q.get()] = idx
+    return [(opening, closing) for opening, closing in pairs.items() if closing]
```

Comment:
Maybe rather use `... if closing is not None` as this is the precise check for "unknown closing position".

Reply:
BTW, maybe that sounds nitpicky, but I recently shot myself in the foot with `if repository:` instead of `if repository is not None`. That can behave rather differently than expected if the repository object happens to have a `__len__` method...

Reply:
Oh that's right! Does not sound nitpicky to me at all. I've added the change and squashed the last couple commits as fixup commits.

Thank you for shepherding this through. I'm proud to have contributed a little piece to such awesome software : )